File: polished-manual-app-builder/services/collab-engine/src/error-capture.ts
import fs from 'fs/promises';
import path from 'path';
import { createHash } from 'crypto';

interface ErrorLog {
  timestamp: string;
  service: string;
  context: string;
  error_type: string;
  error_message: string;
  severity: string;
  stack_trace?: string;
  additional_data?: any;
  environment: string;
  version: string;
  request?: {
    method?: string;
    url?: string;
    headers?: any;
    user_agent?: string;
  };
}

interface ErrorPattern {
  pattern_hash: string;
  error_type: string;
  service: string;
  context: string;
  message: string;
  first_seen: string;
  last_seen: string;
  occurrence_count: number;
  resolved: boolean;
  solution?: string;
  prevention_tips: string[];
  related_errors: string[];
  severity: string;
}

export class ErrorCapture {
  private logDir: string;
  private serviceName: string;

  constructor() {
    this.logDir = path.join(__dirname, '../../logs/errors');
    this.serviceName = 'collab-engine';
    this.ensureLogDirectory();
  }

  private async ensureLogDirectory(): Promise<void> {
    try {
      await fs.mkdir(this.logDir, { recursive: true });
    } catch (error) {
      console.error('Failed to create log directory:', error);
    }
  }

  async logError(
    error: Error,
    context: string,
    request?: any,
    additionalData?: any
  ): Promise<void> {
    const errorLog: ErrorLog = {
      timestamp: new Date().toISOString(),
      service: this.serviceName,
      context,
      error_type: error.constructor.name,
      error_message: error.message,
      severity: this.determineSeverity(error),
      stack_trace: error.stack,
      additional_data: additionalData || {},
      environment: process.env.NODE_ENV || 'development',
      version: process.env.npm_package_version || '1.0.0',
    };

    // Add request context if available
    if (request) {
      errorLog.request = {
        method: request.method,
        url: request.url,
        headers: request.headers,
        user_agent: request.headers?.['user-agent'],
      };
    }

    // Log to structured error file
    await this.writeErrorLog(errorLog);

    // Send to error monitoring system
    await this.sendToMonitor(errorLog);

    // Update error guide if this is a new pattern
    await this.updateErrorGuide(errorLog);

    // Log to console for immediate visibility
    console.error(`[ERROR_CAPTURE] ${context}:`, {
      type: errorLog.error_type,
      severity: errorLog.severity,
      message: error.message,
    });
  }

  private determineSeverity(error: Error): string {
    const errorMessage = error.message.toLowerCase();
    const errorType = error.constructor.name;

    // Critical errors
    if (
      errorType === 'TypeError' ||
      errorType === 'ReferenceError' ||
      errorMessage.includes('out of memory') ||
      errorMessage.includes('segmentation fault')
    ) {
      return 'critical';
    }

    // High severity errors
    if (
      errorType === 'Error' ||
      errorMessage.includes('connection') ||
      errorMessage.includes('timeout') ||
      errorMessage.includes('database') ||
      errorMessage.includes('auth')
    ) {
      return 'high';
    }

    // Medium severity errors
    if (
      errorType === 'ValidationError' ||
      errorMessage.includes('invalid') ||
      errorMessage.includes('not found') ||
      errorMessage.includes('permission')
    ) {
      return 'medium';
    }

    return 'low';
  }

  private async writeErrorLog(errorLog: ErrorLog): Promise<void> {
    try {
      const logFile = path.join(this.logDir, `${this.serviceName}-errors.jsonl`);
      const logLine = JSON.stringify(errorLog) + '\n';
      await fs.appendFile(logFile, logLine);
    } catch (error) {
      console.error('Failed to write error log:', error);
    }
  }

  private async sendToMonitor(errorLog: ErrorLog): Promise<void> {
    try {
      const response = await fetch('http://localhost:8090/errors', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(errorLog),
        signal: AbortSignal.timeout(5000),
      });

      if (!response.ok) {
        console.warn('Failed to send error to monitor:', response.statusText);
      }
    } catch (error) {
      console.warn('Failed to send error to monitor:', error);
    }
  }

  private async updateErrorGuide(errorLog: ErrorLog): Promise<void> {
    try {
      // Create error pattern hash for deduplication
      const patternData = `${errorLog.error_type}${errorLog.error_message}`;
      const patternHash = createHash('md5').update(patternData).digest('hex');

      const newErrorEntry: ErrorPattern = {
        pattern_hash: patternHash,
        error_type: errorLog.error_type,
        service: errorLog.service,
        context: errorLog.context,
        message: errorLog.error_message,
        first_seen: errorLog.timestamp,
        last_seen: errorLog.timestamp,
        occurrence_count: 1,
        resolved: false,
        solution: undefined,
        prevention_tips: [],
        related_errors: [],
        severity: errorLog.severity,
      };

      // Write to pending errors file for review and integration
      const pendingFile = path.join(this.logDir, 'pending-error-patterns.jsonl');
      const entryLine = JSON.stringify(newErrorEntry) + '\n';
      await fs.appendFile(pendingFile, entryLine);
    } catch (error) {
      console.error('Failed to update error guide:', error);
    }
  }
}

// Global error capture instance
export const errorCapture = new ErrorCapture();

// Express middleware for error capture
export function errorCaptureMiddleware(
  error: Error,
  req: any,
  res: any,
  next: any
): void {
  // Capture the error
  errorCapture.logError(
    error,
    `${req.method} ${req.path}`,
    req,
    {
      body: req.body,
      params: req.params,
      query: req.query,
    }
  );

  // Send error response
  const statusCode = (error as any).statusCode || 500;
  res.status(statusCode).json({
    error: 'An error occurred',
    message: process.env.NODE_ENV === 'production' ? 'Internal Server Error' : error.message,
    timestamp: new Date().toISOString(),
  });
}

// WebSocket error capture
export class WebSocketErrorCapture {
  static async captureWebSocketError(
    error: Error,
    context: string,
    socketInfo?: any
  ): Promise<void> {
    await errorCapture.logError(
      error,
      `websocket_${context}`,
      undefined,
      {
        socket_id: socketInfo?.id,
        room: socketInfo?.room,
        user_id: socketInfo?.userId,
        connection_count: socketInfo?.connectionCount,
      }
    );
  }

  static async captureCollaborationError(
    error: Error,
    operation: string,
    documentId: string,
    userId?: string
  ): Promise<void> {
    await errorCapture.logError(
      error,
      `collaboration_${operation}`,
      undefined,
      {
        document_id: documentId,
        user_id: userId,
        operation,
      }
    );
  }

  static async captureEventStoreError(
    error: Error,
    streamId: string,
    eventType: string
  ): Promise<void> {
    await errorCapture.logError(
      error,
      'event_store_operation',
      undefined,
      {
        stream_id: streamId,
        event_type: eventType,
        event_store_url: process.env.EVENT_STORE_URL || 'http://localhost:8080',
      }
    );
  }
}

// Global error handlers
process.on('uncaughtException', async (error: Error) => {
  await errorCapture.logError(error, 'uncaught_exception');
  console.error('Uncaught Exception:', error);
  process.exit(1);
});

process.on('unhandledRejection', async (reason: any) => {
  const error = reason instanceof Error ? reason : new Error(String(reason));
  await errorCapture.logError(error, 'unhandled_rejection');
  console.error('Unhandled Rejection:', reason);
});

export default ErrorCapture;

File: polished-manual-app-builder/services/orchestrator/app/error_capture.py
import json
import traceback
import hashlib
from datetime import datetime
from pathlib import Path
from typing import Optional, Dict, Any
import aiofiles
import httpx
from fastapi import Request, HTTPException
from fastapi.responses import JSONResponse
import structlog

logger = structlog.get_logger()

class ErrorCapture:
    """Captures and analyzes errors for automatic learning and guide updates"""
    
    def __init__(self):
        self.log_dir = Path("../../logs/errors")
        self.log_dir.mkdir(parents=True, exist_ok=True)
        self.service_name = "ai-orchestrator"
        
    async def log_error(
        self, 
        error: Exception, 
        context: str, 
        request: Optional[Request] = None,
        additional_data: Optional[Dict[str, Any]] = None
    ):
        """Log error with full context for analysis and learning"""
        
        error_log = {
            "timestamp": datetime.utcnow().isoformat(),
            "service": self.service_name,
            "context": context,
            "error_type": type(error).__name__,
            "error_message": str(error),
            "severity": self._determine_severity(error),
            "stack_trace": traceback.format_exc(),
            "additional_data": additional_data or {},
            "environment": self._get_environment(),
            "version": "1.0.0",
        }
        
        # Add request context if available
        if request:
            error_log["request"] = {
                "method": request.method,
                "url": str(request.url),
                "headers": dict(request.headers),
                "client_ip": request.client.host if request.client else None,
            }
        
        # Log to structured error file
        await self._write_error_log(error_log)
        
        # Send to error monitoring system
        await self._send_to_monitor(error_log)
        
        # Update error guide if this is a new pattern
        await self._update_error_guide(error_log)
        
        # Log to application logger
        logger.error(
            "Error captured",
            error_type=error_log["error_type"],
            context=context,
            severity=error_log["severity"]
        )
    
    def _determine_severity(self, error: Exception) -> str:
        """Determine error severity based on error type"""
        if isinstance(error, (ConnectionError, TimeoutError)):
            return "high"
        elif isinstance(error, HTTPException):
            if error.status_code >= 500:
                return "high"
            elif error.status_code >= 400:
                return "medium"
            else:
                return "low"
        elif isinstance(error, (ValueError, TypeError)):
            return "medium"
        elif isinstance(error, (FileNotFoundError, PermissionError)):
            return "high"
        else:
            return "medium"
    
    def _get_environment(self) -> str:
        """Get current environment"""
        import os
        return os.getenv("ENVIRONMENT", "development")
    
    async def _write_error_log(self, error_log: Dict[str, Any]):
        """Write error to JSONL log file"""
        log_file = self.log_dir / f"{self.service_name}-errors.jsonl"
        
        try:
            async with aiofiles.open(log_file, "a") as f:
                await f.write(json.dumps(error_log) + "\n")
        except Exception as e:
            logger.error(f"Failed to write error log: {e}")
    
    async def _send_to_monitor(self, error_log: Dict[str, Any]):
        """Send error to monitoring system"""
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                await client.post(
                    "http://localhost:8090/errors",
                    json=error_log
                )
        except Exception as e:
            logger.warning(f"Failed to send error to monitor: {e}")
    
    async def _update_error_guide(self, error_log: Dict[str, Any]):
        """Update error guide with new error patterns"""
        try:
            # Create error pattern hash for deduplication
            pattern_data = f"{error_log['error_type']}{error_log['error_message']}"
            pattern_hash = hashlib.md5(pattern_data.encode()).hexdigest()
            
            new_error_entry = {
                "pattern_hash": pattern_hash,
                "error_type": error_log["error_type"],
                "service": error_log["service"],
                "context": error_log["context"],
                "message": error_log["error_message"],
                "first_seen": error_log["timestamp"],
                "last_seen": error_log["timestamp"],
                "occurrence_count": 1,
                "resolved": False,
                "solution": None,
                "prevention_tips": [],
                "related_errors": [],
                "severity": error_log["severity"]
            }
            
            # Write to pending errors file for review and integration
            pending_file = self.log_dir / "pending-error-patterns.jsonl"
            async with aiofiles.open(pending_file, "a") as f:
                await f.write(json.dumps(new_error_entry) + "\n")
                
        except Exception as e:
            logger.error(f"Failed to update error guide: {e}")


# Global error capture instance
error_capture = ErrorCapture()


async def error_capture_middleware(request: Request, call_next):
    """FastAPI middleware to capture all errors"""
    try:
        response = await call_next(request)
        return response
    except Exception as error:
        # Capture the error
        await error_capture.log_error(
            error=error,
            context=f"{request.method} {request.url.path}",
            request=request
        )
        
        # Return appropriate error response
        if isinstance(error, HTTPException):
            raise error
        else:
            return JSONResponse(
                status_code=500,
                content={
                    "error": "Internal Server Error",
                    "message": "An unexpected error occurred",
                    "timestamp": datetime.utcnow().isoformat()
                }
            )


class ErrorCaptureHandler:
    """Handler for manual error capture in specific contexts"""
    
    @staticmethod
    async def capture_ollama_error(error: Exception, model: str, prompt: str):
        """Capture Ollama-specific errors with context"""
        await error_capture.log_error(
            error=error,
            context="ollama_inference",
            additional_data={
                "model": model,
                "prompt_length": len(prompt),
                "ollama_endpoint": "http://localhost:11434"
            }
        )
    
    @staticmethod
    async def capture_cache_error(error: Exception, operation: str, key: str):
        """Capture cache-related errors"""
        await error_capture.log_error(
            error=error,
            context=f"cache_{operation}",
            additional_data={
                "cache_key": key,
                "operation": operation
            }
        )
    
    @staticmethod
    async def capture_model_error(error: Exception, model_id: str, operation: str):
        """Capture model management errors"""
        await error_capture.log_error(
            error=error,
            context=f"model_{operation}",
            additional_data={
                "model_id": model_id,
                "operation": operation
            }
        )
